internal/contract/infrastructure_machine.go

```diff
@@ -0,0 +1,136 @@
+	})
+	return infrastructureMachine
+}
+
+// Ready provides access to status.ready field in an InfrastructureMachine object.
+func (m *InfrastructureMachineContract) Ready() *Bool {
+	return &Bool{
+		path: []string{"status", "ready"},
+	}
+}
+
+// FailureReason provides access to the status.failureReason field in an InfrastructureMachine object. Note that this field is optional.
+func (m *InfrastructureMachineContract) FailureReason() *String {
+	return &String{
+		path: []string{"status", "failureReason"},
+	}
+}
+
+// FailureMessage provides access to the status.failureMessage field in an InfrastructureMachine object. Note that this field is optional.
+func (m *InfrastructureMachineContract) FailureMessage() *String {
+	return &String{
+		path: []string{"status", "failureMessage"},
+	}
+}
+
+// Addresses provides access to the status.addresses field in an InfrastructureMachine object. Note that this field is optional.
+func (m *InfrastructureMachineContract) Addresses() *MachineAddresses {
+	return &MachineAddresses{
+		path: []string{"status", "addresses"},
+	}
+}
+
+// ProviderID provides access to the spec.providerID field in an InfrastructureMachine object.
+func (m *InfrastructureMachineContract) ProviderID() *String {
+	return &String{
+		path: []string{"spec", "providerID"},
+	}
+
+}
+
+// FailureDomain provides access to the spec.failureDomain field in an InfrastructureMachine object. Note that this field is optional.
+func (m *InfrastructureMachineContract) FailureDomain() *String {
+	return &String{
+		path: []string{"spec", "failureDomain"},
+	}
+}
+
+// MachineAddresses represents an accessor to a []clusterv1.MachineAddress path value.
+type MachineAddresses struct {
+	path Path
+}
+
+// Path returns the path to the []clusterv1.MachineAddress value.
+func (m *MachineAddresses) Path() Path {
+	return m.path
+}
+
+// Get gets the metav1.MachineAddressList value.
+func (m *MachineAddresses) Get(obj *unstructured.Unstructured) (*[]clusterv1.MachineAddress, error) {
+	slice, ok, err := unstructured.NestedSlice(obj.UnstructuredContent(), m.path...)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get %s from object", "."+strings.Join(m.path, "."))
+	}
+	if !ok {
+		return nil, errors.Wrapf(errNotFound, "path %s", "."+strings.Join(m.path, "."))
+	}
+
+	addresses := make([]clusterv1.MachineAddress, len(slice))
+	s, err := json.Marshal(slice)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to marshall field at %s to json", "."+strings.Join(m.path, "."))
+	}
+	json.Unmarshal(s, &addresses)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to unmarshall field at %s to json", "."+strings.Join(m.path, "."))
+	}
+
+	return &addresses, nil
+}
+
+// Set sets the []clusterv1.MachineAddress value in the path.
+func (m *MachineAddresses) Set(obj *unstructured.Unstructured, values []clusterv1.MachineAddress) error {
+	slice := make([]interface{}, len(values))
+	s, err := json.Marshal(values)
```

Comment:
@sbueringer @fabriziopandini One thing I am curious about is if there's a better way to implement the Set() and Get() methods for nested structs. Right now I'm just marshalling the value into a string and unmarshalling it into either the struct or a interface.

Reply:
I don't think so. Basically, we are mimicking what unstructured does; AFIAK also other implementations like duck typing in KNative use a similar approach.

Reply:
Doesn't look like it.

Wondering if we should use `SetNestedSlice` though

Reply:
Gotcha I'll just leave it as is then. Also [SetNestedSlice()](https://github.com/kubernetes/apimachinery/blob/v0.25.2/pkg/apis/meta/v1/unstructured/helpers.go#L244) is implemented just as a wrapper for SetNestedField() so I figured it's all the same.